### PR TITLE
Make CFI evaluation signal safe

### PR DESCRIPTION
**Pull request by fitzgen**
See individual commits for details.

r? @philipc

### CODE REVIEW

File: src/cfi.rs [hunk 1]
```diff
@@ -1271,9 +1282,18 @@ impl<'input, Endian, Section> UninitializedUnwindContext<'input, Endian, Section
 {
     /// Construct a new call frame unwinding context.
     pub fn new() -> UninitializedUnwindContext<'input, Endian, Section> {
-        UninitializedUnwindContext(UnwindContext::new())
+        UninitializedUnwindContext(Box::new(UnwindContext::new()))
```
Comment (philipc) [inline src/cfi.rs]: Is this the only place that allocates now?

**Comment by fitzgen** [comment]
Thanks for the review :)

> Did you measure how much difference the u8 register made? I think it's fine to do anyway, but I suspect it hasn't gained much due to alignment.

I didn't measure in isolation, unfortunately. I was just trying to get everything down to a somewhat reasonable size. I think we may want to make the whole storage generic in the future in case people want more stack size or something like that.

I was dismayed that from a glance that most places wouldn't benefit, but I think a few did.

/me shrugs

> We may be able to reduce RegisterRule size by changing Expression and ValExpression to store only offsets, and keep the common buffer pointer elsewhere.

Nice idea -- I'll file an issue to investigate further.

> FYI, libunwind has its own mempool routines that it uses to allocate the equivalent of UnwindTableRow. It also sets the number of registers based on architecture.

Interesting that they both pre-allocate memory and use mmap. I didn't think that mmap was signal safe, but perhaps I am wrong, or they don't do it during unwinding proper.

But yeah, I think down the line we want to be generic across storage for the register rules, cfa and stack (maybe make `UnwindContext` into a trait?).

File: src/cfi.rs (hunk 1 above)
Comment (fitzgen) [inline src/cfi.rs]: For CFI stuff, yes it should be.

**Comment by philipc** [comment]
> I didn't think that mmap was signal safe, but perhaps I am wrong.

I think that system calls are ok in general because they are atomic from the application's point of view (they return EINTR if interrupted). You have to be sure that the libc function is only a system call and not something more though.

**Pull request merged**